.github/workflows/test.yaml

```diff
@@ -0,0 +1,607 @@
+          mkdir ${LLVM_DIR}
+          tar xf llvm.tar.xz --strip-components=1 -C ${LLVM_DIR}
+          echo "${LLVM_DIR}/bin" >> $GITHUB_PATH
+          echo "LLVM_SYS_120_PREFIX=${LLVM_DIR}" >> $GITHUB_ENV
+        env:
+          LLVM_DIR: .llvm
+      - name: Setup Rust target
+        shell: bash
+        run: |
+          mkdir -p .cargo
+          cat << EOF > .cargo/config.toml
+          [build]
+          target = "${{ matrix.metadata.target }}"
+          EOF
+        if: matrix.metadata.target
+      - name: Cache
+        uses: whywaita/actions-cache-s3@v2
+        with:
+          path: |
+            ~/.cargo/*
+            ./target/*
+          key: s3-${{ github.repository }}-${{ runner.os }}-${{ hashFiles('Cargo.lock') }}-wasmer-make-build-wasmer-${{ matrix.build-what.key }}-${{ matrix.metadata.build }}
+          aws-s3-bucket: wasmer-rust-artifacts-cache
+          aws-access-key-id: ${{ secrets.CLOUDFLARE_ARTIFACTS_CACHE_ACCESS_TOKEN }}
+          aws-secret-access-key: ${{ secrets.CLOUDFLARE_ARTIFACTS_CACHE_ACCESS_KEY }}
+          aws-region: auto
+          aws-endpoint: https://1541b1e8a3fc6ad155ce67ef38899700.r2.cloudflarestorage.com
+          aws-s3-bucket-endpoint: false
+          aws-s3-force-path-style: true
+      - name: Build ${{ matrix.build-what.key }}
+        shell: bash
+        # tests were disabled for linux-musl
+        if: ${{ !(matrix.build-what.key == 'capi' && matrix.metadata.build == 'linux-musl') }}
+        run: |
+          ${{ matrix.build-what.build-cmd }}
+      - name: Build ${{ matrix.build-what.key }}
+        shell: bash
+        if: matrix.build-what.key == 'capi' && matrix.metadata.build == 'linux-musl'
+        run: |
+          make build-capi && make package-capi && make tar-capi
+      - name: Archive production artifacts
+        uses: actions/upload-artifact@v3
+        with:
+          name: wasmer-cli-${{ matrix.metadata.build }}
+          path: build-wasmer.tar.gz
+          if-no-files-found: ignore
+          retention-days: 2
+      - name: Archive production artifacts
+        uses: actions/upload-artifact@v3
+        with:
+          name: capi-${{ matrix.metadata.build }}
+          path: build-capi.tar.gz
+          if-no-files-found: ignore
+          retention-days: 2
+
+  test: 
+    name: ${{ matrix.stage.description }} on ${{ matrix.metadata.build }}
+    runs-on: ${{ matrix.metadata.os }}
+    needs: setup
+    strategy:
+      fail-fast: false
+      matrix:
+        stage: [
+          {
+            stage: 0,
+            description: 'Test compilers', 
+          },
+          {
+            stage: 1, 
+            description: 'cargo test --all',
+          },
+          {
+            stage: 2, 
+            description: 'Test compiler-cranelift',
+          },
+          {
+            stage: 3, 
+            description: 'Test compiler-singlepass',
+          },
+          {
+            stage: 4, 
+            description: 'Test lib/cli',
```

Comment:
As far as I can see from the Makefile, there are three major stages, with some sub-stages: `test packages` (cargo test), `test examples` (cargo test --examples) and `test compilers` (cannot be merged with the other steps).

One of the test also test without WASI, one with WASI.

The most I can merge is these three steps. Let's see if there's any recompilation happening between them (if they use the same feature sets), and otherwise revert the merging.

I can imagine that the CLI itself pulls in a lot more dependencies than just the ones necessary for testing wasmer, so that's probably why it's in a separate step.

Reply:
It does add significant time to the compilation of the tests, I can't merge test stages together. My original solution is still the best one.